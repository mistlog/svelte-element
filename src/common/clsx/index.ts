/**
 * from https://github.com/lukeed/clsx
 */
function toVal(mix) {
	var k, y, str = '';

	if (typeof mix === 'string' || typeof mix === 'number') {
		str += mix;
	} else if (typeof mix === 'object') {
		if (Array.isArray(mix)) {
			for (k = 0; k < mix.length; k++) {
				if (mix[k]) {
					if (y = toVal(mix[k])) {
						str && (str += ' ');
						str += y;
					}
				}
			}
		} else {
			for (k in mix) {
				if (mix[k]) {
					str && (str += ' ');
					str += k;
				}
			}
		}
	}

	return str;
}

export type ClassValue = ClassArray | ClassDictionary | string | number | null | boolean | undefined;

export interface ClassDictionary {
	[id: string]: any;
}

export interface ClassArray extends Array<ClassValue> { }

export function Î£(...args: ClassArray) {
	var i = 0, tmp, x, str = '';
	while (i < arguments.length) {
		if (tmp = arguments[i++]) {
			if (x = toVal(tmp)) {
				str && (str += ' ');
				str += x
			}
		}
	}
	return str;
}
