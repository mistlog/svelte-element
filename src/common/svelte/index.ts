import { setContext as rawSetContext, getContext as rawGetContext } from "svelte";

export function setContext<T>(key: keyof T, value: T[keyof T]) {
    return rawSetContext(key, value);
}

export function getContext<T>(key: keyof T): T[keyof T] {
    return rawGetContext(key);
}

export type SvelteInternalProps = { $$slots?: any };
