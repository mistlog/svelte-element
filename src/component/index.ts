export { default as Button } from "./button/Button";
export * from "./button/Button";

export { default as ButtonGroup } from "./button/ButtonGroup";

export { default as Row } from "./layout/Row";
export { default as Col } from "./layout/Col";

export { default as Radio } from "./radio/Radio";
export * from "./radio/Radio";
export { default as RadioGroup } from "./radio/RadioGroup";
export * from "./radio/RadioGroup";

export { default as Checkbox } from "./checkbox/Checkbox";
export * from "./checkbox/Checkbox";
export { default as CheckboxGroup } from "./checkbox/CheckboxGroup";
export * from "./checkbox/CheckboxGroup";

export { default as Menu } from "./menu/Menu";
export { default as MenuItem } from "./menu/MenuItem";
export { default as SubMenu } from "./menu/SubMenu";
