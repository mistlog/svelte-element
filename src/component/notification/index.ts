import Notification, { INotificationWillUnmountEvent, NotificationType } from "./Notification";

export interface INotificationConfig {
    title: string;
    message: string;
    duration?: number;
    type?: NotificationType;
}

export function notification(config: INotificationConfig) {
    const { title = "", message = "", duration = 3000, type } = config || {};

    //
    const div = document.createElement("div");
    document.body.appendChild(div);

    const className = ".el-notification";

    // credit: element react
    const instances = document.querySelectorAll(className);
    const lastInstance = instances[instances.length - 1] as HTMLElement;
    const top = lastInstance ? lastInstance.getBoundingClientRect().bottom + 16 : 16;

    let component = Reflect.construct(Notification, [
        {
            target: div,
            props: { top, title, message, type },
            intro: true
        }
    ]);

    //
    const removeNotification = () => {
        if (component) {
            component.$destroy();
            component = null;
        }
    };

    component.$on("close", () => {
        removeNotification();
    });

    component.$on("willUnmount", (e: INotificationWillUnmountEvent) => {
        // credit: element react
        const { height, top } = e.detail;
        const instances = document.querySelectorAll(className);
        for (let i = 0; i < instances.length; ++i) {
            const element = instances[i] as HTMLElement;
            const rect = element.getBoundingClientRect();
            if (rect.top > top) {
                element.style.setProperty("top", (rect.top - height - 16).toString() + "px");
            }
        }
        document.body.removeChild(div);
    });

    setTimeout(() => {
        removeNotification();
    }, duration);
}
