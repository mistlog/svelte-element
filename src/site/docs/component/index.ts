import h1 from "./h1";
import h2 from "./h2";
import h3 from "./h3";

export { h1, h2, h3 };