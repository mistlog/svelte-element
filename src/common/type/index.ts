import { Readable, Writable } from "svelte/store";

export type InputOnChangeHandler = (event: Svelte.ChangeEvent<HTMLInputElement>) => void;
export type ButtonOnClickHandler = (event: Svelte.MouseEvent<HTMLButtonElement, MouseEvent>) => void;

type TypeWithGeneric<T> = Readable<T> | Writable<T>;
export type AutoSubscribe<Type> = Type extends TypeWithGeneric<infer X> ? X : never;

export type IconName = "edit" | "share" | "delete" | "search" | "arrow-left";