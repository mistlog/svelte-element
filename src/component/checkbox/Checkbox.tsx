function UI(state: ICheckboxState, { checked }: ICheckboxProps) {
    <label class={state.style}>
        <span class={state.inputWrapperStyle}>
            <span class={state.inputSymbolStyle} />
            <input type="checkbox" checked={checked} class={state.inputStyle} onChange />
        </span>
        <span class={state.inputLabelStyle}>
            <slot />
        </span>
    </label>;
}

export interface ICheckboxProps {
    /**
     * if the Checkbox is checked
     * @default false
     */
    checked?: boolean;
    /**
     * whether the Checkbox is disabled
     * @default false
     */
    disabled?: boolean;
    onChange?: InputOnChangeHandler;
}

export default function _(props: ICheckboxProps) {
    const { checked = false, disabled = false } = props;

    //@ts-ignore
    <State />;

    //@ts-ignore
    <UI />;
}

interface ICheckboxState {
    inputStyle: string;
    inputSymbolStyle: string;
    inputWrapperStyle: string;
    inputLabelStyle: string;
    style: string;
}

function State({ checked, disabled }: ICheckboxProps) {
    const state = {} as ICheckboxState;

    state.inputSymbolStyle = "el-checkbox__inner";
    state.inputStyle = "el-checkbox__original";
    state.inputLabelStyle = "el-checkbox__label";
    state.style = "el-checkbox";

    {
        ("use watch");
        state.inputWrapperStyle = Σ("el-checkbox__input", {
            "is-checked": checked,
            "is-disabled": disabled
        });
    }
}

import { Σ } from "../../common/clsx/index";
import { InputOnChangeHandler } from "../../common/type";
