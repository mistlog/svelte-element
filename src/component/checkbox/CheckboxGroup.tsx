function UI(
    state: ICheckboxGroupState,
    { options }: ICheckboxGroupProps,
    onChangeHandler: (checkbox: ICheckbox) => void,
    checkbox: ICheckbox
) {
    <each from={options}>
        {(checkbox: ICheckbox, key = checkbox.label) => (
            <CheckBox
                disabled={Boolean(checkbox.disabled)}
                checked={state.selected.includes(checkbox.label)}
                onChange={event => onChangeHandler(checkbox)}
            >
                {checkbox.label}
            </CheckBox>
        )}
    </each>;
}

export interface ICheckbox {
    label: string;
    disabled?: boolean;
}

export interface ICheckboxGroupProps {
    options?: Array<ICheckbox>;
    defaultSelected?: Array<string>;
    onChange?: CheckboxGroupOnChangeHandler;
}

export default function _(props: ICheckboxGroupProps) {
    const { options, defaultSelected = [] } = props;

    //@ts-ignore
    <State />;

    //@ts-ignore
    <Event />;

    //@ts-ignore
    <UI />;
}

//
interface ICheckboxGroupState {
    selected: Array<string>;
}

function State({ defaultSelected }: ICheckboxGroupProps) {
    const state: ICheckboxGroupState = { selected: defaultSelected };
}

//
export type CheckboxGroupEvent = { detail: { selected: Array<string> } };
export type CheckboxGroupOnChangeHandler = (event: CheckboxGroupEvent) => void;

function Event(state: ICheckboxGroupState) {
    const dispatch = createEventDispatcher();
    const onChangeHandler = (checkbox: ICheckbox) => {
        const { label } = checkbox;
        if (state.selected.includes(label)) {
            state.selected = state.selected.filter(item => item !== label);
        } else {
            state.selected = [label, ...state.selected];
        }
        dispatch("change", { selected: state.selected });
    };
}

import { createEventDispatcher } from "svelte";
import { default as CheckBox } from "./Checkbox";
