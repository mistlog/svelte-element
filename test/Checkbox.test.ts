import { render, fireEvent } from "@testing-library/svelte";
import { normalize } from "./common/utility";

import { ICheckboxProps, ICheckboxGroupProps } from "../src/component";
import Checkbox from "../src/component/checkbox/Checkbox.svelte";
import CheckboxGroup from "../src/component/checkbox/CheckboxGroup.svelte";

import DefaultCheckbox from "./assets/DefaultCheckbox.svelte";
import ClickCheckbox from "./assets/ClickCheckbox.svelte";
import ThreeCheckboxes from "./assets/ThreeCheckboxes.svelte";

describe("Checkbox", () => {
    test("text", () => {
        const { container } = render(DefaultCheckbox);
        const checkbox = container.querySelector(".el-checkbox");
        expect(normalize(checkbox.outerHTML)).toMatchSnapshot();
    });

    test("checked", () => {
        const { container } = render(Checkbox, { checked: true } as ICheckboxProps);
        const input = container.querySelector("input");
        const wrapper = input.parentElement as HTMLSpanElement;
        expect(wrapper.className.includes("is-checked")).toEqual(true);
    });

    test("disabled", () => {
        const { container } = render(Checkbox, { disabled: true } as ICheckboxProps);
        const input = container.querySelector("input");
        const wrapper = input.parentElement as HTMLSpanElement;
        expect(wrapper.className.includes("is-disabled")).toEqual(true);
    });

    test("click event", async () => {
        const { container } = render(ClickCheckbox);
        const input = container.querySelector("input");
        expect(input.checked).toEqual(true);
        await fireEvent(input, new MouseEvent("click", {}));
        expect(input.checked).toEqual(false);
    });
});

describe("CheckboxGroup", () => {
    test("empty checkbox group", () => {
        const { container } = render(CheckboxGroup, { options: [] } as ICheckboxGroupProps);
        expect(normalize(container.outerHTML)).toMatchSnapshot();
    });

    test("check group with checkboxes", async () => {
        const { container } = render(ThreeCheckboxes);
        expect(normalize(container.outerHTML)).toMatchSnapshot();

        const [a, b, c] = container.getElementsByTagName("input");
        await fireEvent(a, new MouseEvent("click", {}));
        await fireEvent(b, new MouseEvent("click", {}));
        await fireEvent(c, new MouseEvent("click", {}));
        expect(normalize(container.outerHTML)).toMatchSnapshot();
    });
});
