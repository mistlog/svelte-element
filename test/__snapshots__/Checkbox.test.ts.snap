// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox text 1`] = `
"<label class=\\"el-checkbox\\">
  <span class=\\"el-checkbox__input\\">
    <span class=\\"el-checkbox__inner\\"></span>
    <input type=\\"checkbox\\" class=\\"el-checkbox__original\\" />
  </span>
  <span class=\\"el-checkbox__label\\">Option</span>
</label>
"
`;

exports[`CheckboxGroup check group with checkboxes 1`] = `
"<body>
  <div>
    <label class=\\"el-checkbox\\">
      <span class=\\"el-checkbox__input is-checked\\">
        <span class=\\"el-checkbox__inner\\"></span>
        <input type=\\"checkbox\\" class=\\"el-checkbox__original\\" />
      </span>
      <span class=\\"el-checkbox__label\\">Option A</span>
    </label>
    <label class=\\"el-checkbox\\">
      <span class=\\"el-checkbox__input\\">
        <span class=\\"el-checkbox__inner\\"></span>
        <input type=\\"checkbox\\" class=\\"el-checkbox__original\\" />
      </span>
      <span class=\\"el-checkbox__label\\">Option B</span>
    </label>
    <label class=\\"el-checkbox\\">
      <span class=\\"el-checkbox__input\\">
        <span class=\\"el-checkbox__inner\\"></span>
        <input type=\\"checkbox\\" class=\\"el-checkbox__original\\" />
      </span>
      <span class=\\"el-checkbox__label\\">Option C</span>
    </label>
  </div>
</body>
"
`;

exports[`CheckboxGroup check group with checkboxes 2`] = `
"<body>
  <div>
    <label class=\\"el-checkbox\\">
      <span class=\\"el-checkbox__input\\">
        <span class=\\"el-checkbox__inner\\"></span>
        <input type=\\"checkbox\\" class=\\"el-checkbox__original\\" />
      </span>
      <span class=\\"el-checkbox__label\\">Option A</span>
    </label>
    <label class=\\"el-checkbox\\">
      <span class=\\"el-checkbox__input is-checked\\">
        <span class=\\"el-checkbox__inner\\"></span>
        <input type=\\"checkbox\\" class=\\"el-checkbox__original\\" />
      </span>
      <span class=\\"el-checkbox__label\\">Option B</span>
    </label>
    <label class=\\"el-checkbox\\">
      <span class=\\"el-checkbox__input is-checked\\">
        <span class=\\"el-checkbox__inner\\"></span>
        <input type=\\"checkbox\\" class=\\"el-checkbox__original\\" />
      </span>
      <span class=\\"el-checkbox__label\\">Option C</span>
    </label>
  </div>
</body>
"
`;

exports[`CheckboxGroup empty checkbox group 1`] = `
"<body>
  <div></div>
</body>
"
`;
